<!-- Marketing messaging and featurettes
    ================================================== -->
<!-- Wrap the rest of the page in another container to center all the content. -->

<div class="container marketing">

    <!-- Three columns of text below the carousel -->
    <div class="row">
        <div class="col-lg-4">
            <img class="img-rounded" src="/tsne-sample.jpg" width="140" height="140">
            <h2>Explore Cell Types</h2>
            <p class=expo>The expression profile for each cell was grouped into clusters and sub-clusters through a semi-automated process and the results were projected onto 2-D t-SNE plots to visually approximate the expression relationships amongst individual cells. Gene expression levels of query genes can be overlayed on any subset of data.</p>
            <p class=expo>Example: <a href="/?_state_id_=629172982ecf1bb2">The relative expression profile of Sepp1 among sub-types of oligodendrocytes in frontal cortex and hippocampus</a></p>
            <p class=buttonpar>
                <1>
            </p>
        </div>
        <!-- /.col-lg-4 -->
        <div class="col-lg-4">
            <img class="img-rounded" src="/scatter-sample.jpg" width="140" height="140">
            <h2>Compare and Discover Genes</h2>
            <p class=expo>The relative pairwise gene expression between cell types within a region or across different regions can be quickly computed for 10,000s of genes and ranked to identify distinguishing marker genes. Additional user selected genes can be added to the comparison set.</p>
            <p class=expo>Example: <a href="/?_state_id_=2b787103224f8802">The set of genes overexpressed at least five-fold in Oligodendrocytes versus Polydendrocytes within the Entopeduncular.</a></p>
            <p class=buttonpar>
                <2>
            </p>
        </div>
        <!-- /.col-lg-4 -->
        <div class="col-lg-4">
            <img class="img-rounded" src="/rank-sample.jpg" width="140" height="140">
            <h2>Search By Gene</h2>
            <p class=expo>Enter one or more genes of interest to immediately see the relative gene expression across 262 cell types.</p>
            <p class=expo>Example: <a href="/?_state_id_=f8677f6733c2fe52">The twenty cell types with the highest expression of S100a6 across all brain regions.</a></p>
            <p class=buttonpar>
                <3>
            </p>
        </div>
        <!-- /.col-lg-4 -->
    </div>
    <!-- /.row -->

      <p class="note">
	We hope you find this data resource useful.  Please <a href="mailto:mouse.dropviz@gmail.com">write us</a> with experiences and suggestions.  We would love stories about how people are using it in their work.</p>

    <!-- START THE FEATURETTES -->

    <hr class="featurette-divider">

    <div class="row featurette">
        <div class="col-md-2"></div>
        <div class="col-md-8">
            <h2 class="featurette-heading">Drop-seq<span class="text-muted"></span></h2>
	    <p><a href="http://mccarrolllab.com/dropseq/">Drop-seq</a> is a technology we developed for highly parallel analysis of RNA expression in thousands of individual cells.  Drop-seq works by encapsulating individual cells into vast numbers of nanoliter-sized droplets, together with DNA-barcoded beads that uniquely identify the droplets.  We described Drop-seq in <a href="http://www.cell.com/abstract/S0092-8674(15)00549-8">a paper in 2015</a>, and subsequent commercial technologies have been built on this approach. We have made Drop-seq open-source, and hundreds of labs have adopted it; detailed protocols and software are available on the <a href="http://mccarrolllab.com/dropseq/">Drop-seq web site</a>.</p>
            <h2 class="featurette-heading">Summary</h2>
	  <p>
	    The mammalian brain is composed of diverse, specialized
	    cell populations. To systematically ascertain and learn
	    from these cellular specializations, we used Drop-seq to
	    profile RNA expression in 690,000 individual cells sampled
	    from 9 regions of the adult mouse brain. We identified 565
	    transcriptionally distinct groups of cells using
	    computational approaches developed to distinguish
	    biological from technical signals. Cross-region analysis
	    of these 565 cell populations revealed features of brain
	    organization, including a gene-expression module for
	    synthesizing axonal and presynaptic components, patterns
	    in the co-deployment of voltage-gated ion channels,
	    functional distinctions among the cells of the vasculature
	    and specialization of glutamatergic neurons across
	    cortical regions. Systematic neuronal classifications for
	    two complex basal ganglia nuclei and the striatum revealed
	    a rare population of spiny projection neurons. This adult
	    mouse brain cell atlas, accessible through interactive
	    online software (DropViz), serves as a reference for
	    development, disease, and evolution.
</p>
            <div style="display:none" class=citation>
                <dl><dt>1</dt>Macosko, E. Z. <em>et al.</em>Cell, Volume 161, Issue 5, pp 1202&ndash;1214 (2015).</dl><dt>2</dt>Gierahn, T. M. <em>et al.</em> Seq-Well: portable, low-cost RNA sequencing of single cells at high throughput. Nature methods 14, pp 395&ndash;398, doi:10.1038/nmeth.4179 (2017).</dl>
            </div>
            <h2 class="featurette-heading">Citation</h2>
            <p>
	      Saunders A*, Macosko E.Z*, Wysoker A, Goldman M, Krienen, F, de Rivera H, Bien E, Baum M, Wang S, Bortolin L, Goeva A, Nemesh J, Kamitaki N, Brumbaugh S, Kulp D and McCarroll, S.A. 2018. Molecular Diversity and Specializations among the Cells of the Adult Mouse Brain. 2018. Cell. 174(4) P1015-1030.E16 (<a href="https://doi.org/10.1016/j.cell.2018.07.028">DOI</a>)
            </p>

        </div>
    </div>
    <div class="col-md-2"></div>

    <hr class="featurette-divider">

    <div class="funding-section">
        <div class="funding">
            <h2 class="featurette-heading">Funding and Support<span class="text-muted"></span></h2>
          <p class="lead">This research was performed in the <a href="http://mccarrolllab.com/">McCarroll</a> and <a href="http://macoskolab.com/">Macosko</a> Labs at Harvard Medical School and the Broad Institute's <a href="https://www.broadinstitute.org/stanley">Stanley Center for Psychiatric Research</a>.</p>
        </div>
          <img class="featurette-image img-responsive center-block" src="/StanleyCenter-web.png" alt="Stanley Center">
          <img width="400" class="featurette-image img-responsive center-block" src="HMS.png" alt="Harvard Medical School">
</div>

    <hr class="featurette-divider">


    <!-- /END THE FEATURETTES -->

</div>

<!-- /.container -->
